File: Android/livestreamcore/src/main/java/com/trtc/uikit/livekit/livestreamcore/LiveCoreViewDefine.java
package com.trtc.uikit.livekit.livestreamcore;

import android.graphics.Rect;
import android.view.View;

import com.tencent.cloud.tuikit.engine.common.TUICommonDefine;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleInfo;
import com.tencent.cloud.tuikit.engine.extension.TUILiveBattleManager.BattleUser;
import com.tencent.cloud.tuikit.engine.extension.TUILiveConnectionManager.ConnectionUser;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine.UserInfo;

import java.util.List;

public class LiveCoreViewDefine {

    public enum LayoutMode {
        GRID_LAYOUT, FLOAT_LAYOUT, FREE_LAYOUT
    }

    public interface ConnectionObserver {
        void onConnectedUsersUpdated(List<UserInfo> userList, List<UserInfo> joinList, List<UserInfo> leaveList);

        void onUserConnectionRequest(UserInfo inviterUser);

        void onUserConnectionCancelled(UserInfo inviterUser);

        void onUserConnectionAccepted(UserInfo userInfo);

        void onUserConnectionRejected(UserInfo userInfo);

        void onUserConnectionTimeout(UserInfo userInfo);

        void onUserConnectionTerminated();

        void onUserConnectionExited(UserInfo userInfo);

        void onConnectedRoomsUpdated(List<ConnectionUser> roomList);

        void onCrossRoomConnectionRequest(ConnectionUser roomInfo);

        void onCrossRoomConnectionCancelled(ConnectionUser roomInfo);

        void onCrossRoomConnectionAccepted(ConnectionUser roomInfo);

        void onCrossRoomConnectionRejected(ConnectionUser roomInfo);

        void onCrossRoomConnectionTimeout(ConnectionUser inviter, ConnectionUser invitee);

        void onCrossRoomConnectionExited(ConnectionUser roomInfo);

        void onRoomDismissed(String roomId);
    }

    public interface BattleObserver {
        void onBattleStarted(BattleInfo battleInfo);

        void onBattleEnded(BattleInfo battleInfo);

        void onUserJoinBattle(String battleId, BattleUser battleUser);

        void onUserExitBattle(String battleId, BattleUser battleUser);

        void onBattleScoreChanged(String battleId, List<BattleUser> battleUserList);

        void onBattleRequestReceived(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestCancelled(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestTimeout(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestAccept(String battleId, BattleUser inviter, BattleUser invitee);

        void onBattleRequestReject(String battleId, BattleUser inviter, BattleUser invitee);
    }

    public interface BattleRequestCallback {
        void onSuccess(String battleId, List<String> requestedUserIdList);

        void onError(TUICommonDefine.Error error, String message);
    }


    public interface VideoViewAdapter {
        View createCoGuestView(TUIRoomDefine.UserInfo userInfo);

        void updateCoGuestView(View coGuestView, TUIRoomDefine.UserInfo userInfo, List<UserInfoModifyFlag> modifyFlag);

        View createCoHostView(CoHostUser coHostUser);

        void updateCoHostView(View coHostView, CoHostUser coHostUser, List<UserInfoModifyFlag> modifyFlag);

        View createBattleView(BattleUser battleUser);

        void updateBattleView(View battleView, BattleUser battleUser);

        View createBattleContainerView();

        void updateBattleContainerView(View battleContainnerView, List<BattleUserViewModel> userInfos);
    }

    public static class BattleUserViewModel {
        public BattleUser battleUser;
        public Rect       rect = new Rect();
    }

    public static class CoHostUser {
        public ConnectionUser connectionUser;
        public boolean        hasAudioStream;
        public boolean        hasVideoStream;

    }

    public enum UserInfoModifyFlag {
        NONE(0x00),
        USER_ROLE(0x01),
        NAME_CARD(0x01 << 1),
        HAS_VIDEO_STREAM(0x01 << 2),
        HAS_AUDIO_STREAM(0x01 << 3);
        final int mValue;

        UserInfoModifyFlag(int value) {
            mValue = value;
        }

        public int getValue() {
            return mValue;
        }

        public static UserInfoModifyFlag fromInt(int value) {
            for (UserInfoModifyFlag flag : UserInfoModifyFlag.values()) {
                if (flag.mValue == value) {
                    return flag;
                }
            }
            return NONE;
        }
    }
}

File: iOS/LiveStreamCore/Sources/LiveCoreViewDefine.swift
//
//  LiveCoreViewDefine.swift
//  LiveStreamCore
//
//  Created by jeremiawang on 2024/10/22.
//

import Foundation
import RTCRoomEngine

@objc public enum LayoutMode: Int {
    case gridLayout
    case floatLayout
    case freeLayout
}

/**************************************** observer ****************************************/
@objc public protocol ConnectionObserver {
    func onConnectedUsersUpdated(userList: [TUIUserInfo], joinList: [TUIUserInfo], leaveList: [TUIUserInfo])
    func onUserConnectionRequest(inviterUser: TUIUserInfo)
    func onUserConnectionCancelled(inviterUser: TUIUserInfo)
    func onUserConnectionAccepted(userInfo: TUIUserInfo)
    func onUserConnectionRejected(userInfo: TUIUserInfo)
    func onUserConnectionTimeout(userInfo: TUIUserInfo)
    func onUserConnectionTerminated()
    func onUserConnectionExited(userInfo: TUIUserInfo)
    func onConnectedRoomsUpdated(hostUserList: [TUIConnectionUser])
    func onCrossRoomConnectionRequest(hostUser: TUIConnectionUser)
    func onCrossRoomConnectionCancelled(hostUser: TUIConnectionUser)
    func onCrossRoomConnectionAccepted(hostUser: TUIConnectionUser)
    func onCrossRoomConnectionRejected(hostUser: TUIConnectionUser)
    func onCrossRoomConnectionTimeout(inviter: TUIConnectionUser, invitee: TUIConnectionUser)
    func onCrossRoomConnectionExited(hostUser: TUIConnectionUser)
    func onRoomDismissed(roomId: String)
}

@objc public protocol BattleObserver {
    func onBattleStarted(battleInfo: TUIBattleInfo)
    func onBattleEnded(battleInfo: TUIBattleInfo)
    func onUserJoinBattle(battleId: String, battleUser: TUIBattleUser)
    func onUserExitBattle(battleId: String, battleUser: TUIBattleUser)
    func onBattleScoreChanged(battleId: String, battleUserList: [TUIBattleUser])
    func onBattleRequestReceived(battleId: String, inviter: TUIBattleUser, invitee: TUIBattleUser)
    func onBattleRequestCancelled(battleId: String, inviter: TUIBattleUser, invitee: TUIBattleUser)
    func onBattleRequestTimeout(battleId: String, inviter: TUIBattleUser, invitee: TUIBattleUser)
    func onBattleRequestAccept(battleId: String, inviter: TUIBattleUser, invitee: TUIBattleUser)
    func onBattleRequestReject(battleId: String, inviter: TUIBattleUser, invitee: TUIBattleUser)
}

@objc public protocol VideoViewDelegate {
    func createCoGuestView(userInfo: TUIUserInfo) -> UIView?
    func updateCoGuestView(coGuestView: UIView, userInfo: TUIUserInfo, modifyFlag: UserInfoModifyFlag)
    func createCoHostView(coHostUser: CoHostUser) -> UIView?
    func updateCoHostView(coHostView: UIView, coHostUser: CoHostUser, modifyFlag: UserInfoModifyFlag)
    func createBattleView(battleUser: TUIBattleUser) -> UIView?
    func updateBattleView(battleView: UIView, battleUser: TUIBattleUser)
    func createBattleContainerView() -> UIView?
    func updateBattleContainerView(battleContainerView: UIView, userInfos: [BattleUserViewModel])
}

@objc public protocol WaitingCoGuestViewDelegate {
    func waitingCoGuestView() -> UIView?
}

@objcMembers public class CoHostUser: NSObject {
    public var connectionUser: TUIConnectionUser = TUIConnectionUser()
    public var hasAudioStream: Bool = false
    public var hasVideoStream: Bool = true
}

@objcMembers public class BattleUserViewModel: NSObject {
    public var battleUser: TUIBattleUser = TUIBattleUser()
    public var rect: CGRect = .zero
}

@objc public class UserInfoModifyFlag: NSObject, OptionSet {
    public let rawValue: UInt

    @objc public static let none = UserInfoModifyFlag([])
    @objc public static let userRole = UserInfoModifyFlag(rawValue: 0x01 << 0)
    @objc public static let nameCard = UserInfoModifyFlag(rawValue: 0x01 << 1)
    @objc public static let hasVideoStream = UserInfoModifyFlag(rawValue: 0x01 << 2)
    @objc public static let hasAudioStream = UserInfoModifyFlag(rawValue: 0x01 << 3)

    required public init(rawValue: UInt) {
        self.rawValue = rawValue
    }
}

public typealias TUIBattleRequestBlock = (String, [TUIBattleUser]) -> Void

// MARK: ------------- Private Notification -------------
// MARK: - Room
public let LiveCoreViewOnEnterRoomNotifyName: Notification.Name = Notification.Name("__kLiveCoreView_NotifyName_Room_EnterRoom__")
public let LiveCoreViewOnExitRoomNotifyName: Notification.Name = Notification.Name("__kLiveCoreView_NotifyName_Room_ExitRoom__")
// MARK: - Video
// MARK: - Audio
